kv/kvserver: TestLeaseRequestFromExpirationToEpochOrLeaderDoesNotRegressExpiration failed

kv/kvserver.TestLeaseRequestFromExpirationToEpochOrLeaderDoesNotRegressExpiration [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18008028?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18008028?buildTab=artifacts#/) on master @ [b3fec61ca90095c664f6432af864f18e9946f8bb](https://github.com/cockroachdb/cockroach/commits/b3fec61ca90095c664f6432af864f18e9946f8bb):


```
=== RUN   TestLeaseRequestFromExpirationToEpochOrLeaderDoesNotRegressExpiration
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/f53f424fd3523210e5809be9ed06c425/logTestLeaseRequestFromExpirationToEpochOrLeaderDoesNotRegressExpiration3845698661
    test_log_scope.go:76: use -show-logs to present logs inline
    client_lease_test.go:1860: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/f53f424fd3523210e5809be9ed06c425/logTestLeaseRequestFromExpirationToEpochOrLeaderDoesNotRegressExpiration3845698661
--- FAIL: TestLeaseRequestFromExpirationToEpochOrLeaderDoesNotRegressExpiration (51.43s)
=== RUN   TestLeaseRequestFromExpirationToEpochOrLeaderDoesNotRegressExpiration/lease-type=LeaseEpoch
    testcluster.go:1137: condition failed to evaluate within 45s: from testcluster.go:1142: lease still an expiration based lease
    --- FAIL: TestLeaseRequestFromExpirationToEpochOrLeaderDoesNotRegressExpiration/lease-type=LeaseEpoch (48.20s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLeaseRequestFromExpirationToEpochOrLeaderDoesNotRegressExpiration.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-45117

kv/kvserver.TestLeaseRequestFromExpirationToEpochOrLeaderDoesNotRegressExpiration [failed](https://tanzanite.cluster.engflow.com/invocations/default/443b875b-c977-4fe3-abd3-6ff428b260d5?testReportRun=23&testReportShard=32&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [d7ea85402dc35e36c6cc35520fa91f25fd5c999d](https://github.com/cockroachdb/cockroach/commits/d7ea85402dc35e36c6cc35520fa91f25fd5c999d):


```
=== RUN   TestLeaseRequestFromExpirationToEpochOrLeaderDoesNotRegressExpiration
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestLeaseRequestFromExpirationToEpochOrLeaderDoesNotRegressExpiration3650053752
    test_log_scope.go:76: use -show-logs to present logs inline
    client_lease_test.go:1877: -- test log scope end --
test logs left over in: outputs.zip/logTestLeaseRequestFromExpirationToEpochOrLeaderDoesNotRegressExpiration3650053752
--- FAIL: TestLeaseRequestFromExpirationToEpochOrLeaderDoesNotRegressExpiration (52.93s)
=== RUN   TestLeaseRequestFromExpirationToEpochOrLeaderDoesNotRegressExpiration/lease-type=LeaseEpoch
    testcluster.go:1137: condition failed to evaluate within 45s: from testcluster.go:1142: lease still an expiration based lease
    --- FAIL: TestLeaseRequestFromExpirationToEpochOrLeaderDoesNotRegressExpiration/lease-type=LeaseEpoch (48.79s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=23</code>
 - <code>shard=32</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLeaseRequestFromExpirationToEpochOrLeaderDoesNotRegressExpiration.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/kv/kvserver/client_lease_test.go
+++ b/pkg/kv/kvserver/client_lease_test.go
@@ -1811,6 +1811,23 @@ func TestLeaseRequestFromExpirationToEpochOrLeaderDoesNotRegressExpiration(t *te
 				// Pause n1's node liveness heartbeats, to allow its liveness expiration to
 				// fall behind.
 				l0 := tc.Server(0).NodeLiveness().(*liveness.NodeLiveness)
+
+				// Wait for the node liveness record to have an epoch set. Otherwise,
+				// the test might fail later when attempting to acquire an epoch-based
+				// lease with this error: unable to get liveness record. This can happen
+				// if we pause the node liveness heartbeat before the record is even
+				// created.
+				testutils.SucceedsSoon(t, func() error {
+					l, ok := l0.GetLiveness(tc.Server(0).NodeID())
+					if !ok {
+						return errors.New("liveness not found")
+					}
+					if l.Epoch == 0 {
+						return errors.New("liveness epoch not set")
+					}
+					return nil
+				})
+
 				l0.PauseHeartbeatLoopForTest()
 				l, ok := l0.GetLiveness(tc.Server(0).NodeID())
 				require.True(t, ok)

